rust/metering/src/lib.rs

```diff
@@ -0,0 +1,334 @@
+extern crate proc_macro;
+
+use proc_macro2::TokenStream;
+use quote::quote;
+
+use crate::{
+    attributes::generate_attribute_definition_token_stream,
+    events::generate_event_definition_token_stream,
+    mutators::generate_noop_mutator_definition_token_stream,
+    utils::{generate_compile_error, process_token_stream},
+};
+
+mod annotations;
+mod attributes;
+mod errors;
+mod events;
+mod fields;
+mod mutators;
+mod utils;
+
+/// This is the only user-facing export of `chroma_metering`. It is responsible for registering attributes and
+/// events by producing the code necessary to allow users to interact with the metering library.
+#[proc_macro]
+pub fn initialize_metering(raw_token_stream: proc_macro::TokenStream) -> proc_macro::TokenStream {
+    let token_stream = TokenStream::from(raw_token_stream);
+
+    let (attributes, events) = match process_token_stream(&token_stream) {
+        Ok(result) => result,
+        Err(error) => return generate_compile_error(&error.to_string()),
+    };
+
+    let noop_mutator_definition_token_streams: Vec<TokenStream> = attributes
+        .iter()
+        .map(generate_noop_mutator_definition_token_stream)
+        .collect();
+
+    let attribute_definition_token_streams: Vec<TokenStream> = attributes
+        .iter()
+        .map(generate_attribute_definition_token_stream)
+        .collect();
+
+    let event_definition_token_streams: Vec<TokenStream> = events
+        .iter()
+        .map(generate_event_definition_token_stream)
+        .collect();
+
+    proc_macro::TokenStream::from(quote! {
+        /// The primary trait used in the metering library that contains no-op mutators for every attribute.
+        pub trait MeteringEvent: std::fmt::Debug + std::any::Any + Send + 'static {
+            #( #noop_mutator_definition_token_streams )*
+        }
+
+        #( #attribute_definition_token_streams )*
+
+        #( #event_definition_token_streams )*
+
+        /// The default receiver registered in the library.
+        #[derive(Clone, std::fmt::Debug)]
+
+        pub struct __DefaultReceiver;
+
+        /// The default receiver simply prints out the metering events submitted to it.
+        #[async_trait::async_trait]
+        impl chroma_system::ReceiverForMessage<Box<dyn MeteringEvent>>
```

Comment:
You don't use the System crate by implementing Receiver, just implement Handler<M>. Receiver is given for free